### PR TITLE
Fix a KaTeX crash and many docstring issues

### LINKED ISSUE #43667: Kwargs are not correctly documented for some operators
Keyword arguments should be put into a separate section `Keyword Arguments` instead of inside `Parameters`, just like:
image
instead of
image

I have three PRs fixing this issue: https://github.com/pytorch/pytorch/pull/43583 https://github.com/pytorch/pytorch/pull/43586 https://github.com/pytorch/pytorch/pull/43589

But some of them can not be fixed due to https://github.com/pytorch/pytorch/issues/43669: List of ops remaining to be fixed:
- `torch.eye`
- `torch.bmm`
- `torch.gather`
- `torch.linspace`
- `torch.logspace`
- `torch.ones`
- `torch.ones_like`
- The entire https://github.com/pytorch/pytorch/pull/43589

cc @jlin27 @mruberry

### PR DESCRIPTION
**Pull request by rgommers**
The first commit fixes the `MultiheadAttention` docstrings, which are causing a cryptic KaTeX crash.

The second commit fixes many documentation issues in `torch/_torch_docs.py`, and closes gh-43667 (missing "Keyword arguments" headers). It also fixes a weird duplicate docstring for `torch.argmin`; there's more of these, it looks like they were written based on whether the C++ implementation has an overload. That makes little sense to a Python user though, and the content is simply duplicate.

The `Shape:` heading for https://pytorch.org/docs/master/generated/torch.nn.MultiheadAttention.html looked bad, here's what it looks like with this PR:

image

### CODE REVIEW

**Comment by rgommers** [comment]
Note: it looks like the `Keyword Arguments` in some cases was/is meant as "this is a keyword-**only** argument". That's a weird thing to separate out, and is already indicated by the `*` in the function signatures. So I interpreted "keyword arguments" as what it says - all keyword args.

**Comment by rgommers** [comment]
For reference/searchability, the relevant part of the KaTeX crash is:
```
/home/rgommers/.npm-global/lib/node_modules/katex/dist/katex.js:17167         
    throw error;
    ^

ParseError: KaTeX parse error: Expected '}', got '_' at position 13: (N*\text{num_̲heads}, L, S)
    at new ParseError (/home/rgommers/.npm-global/lib/node_modules/katex/dist/katex.js:265:14)
```

File: torch/_torch_docs.py [hunk 1]
```diff
@@ -2906,6 +2904,8 @@ def merge_dicts(*dicts):
 
 Args:
     n (int): the number of rows
+
+Keyword arguments:
     m (int, optional): the number of columns with default being :attr:`n`
```
Comment (mruberry) [inline torch/_torch_docs.py:2909]: The convention is that "Keyword arguments" are the keyword-only arguments. See torch.sub, for example:

https://pytorch.org/docs/master/generated/torch.sub.html?highlight=sub#torch.sub

Unrelated to this PR, of course, but I wish NumPy and PyTorch had been much more aggressive about making arguments keyword-only.

File: torch/_torch_docs.py [hunk 2]
```diff
@@ -4489,7 +4489,9 @@ def merge_dicts(*dicts):
     start (float): the starting value for the set of points
     end (float): the ending value for the set of points
     steps (int): size of the constructed tensor
-    base (float): base of the logarithm function. Default: ``10.0``.
+
+Keyword arguments:
+    base (float, optional): base of the logarithm function. Default: ``10.0``.
```
Comment (mruberry) [inline torch/_torch_docs.py:4494]: "base" would also be an "Arg" and not a "Kwarg"

File: torch/_torch_docs.py (hunk 1 above)
Comment (rgommers) [reply torch/_torch_docs.py:2909]: Should we rename "Keyword arguments" then? The current name is plain wrong if it's keyword-only.

> Unrelated to this PR, of course, but I wish NumPy and PyTorch had been much more aggressive about making arguments keyword-only.

working on that: https://data-apis.github.io/array-api/latest/API_specification/function_and_method_signatures.html :)

Comment (rgommers) [reply torch/_torch_docs.py:2909]: (that rename would be my second option), as I commented, separating out keyword-only just doesn't make much sense. Keywords like `dim` and `dtype` are _keywords_, and should preferably be used as such rather than as positional arguments.

File: torch/_torch_docs.py [hunk 3]
```diff
@@ -5390,9 +5392,9 @@ def merge_dicts(*dicts):
 
 add_docstr(torch.argmin,
            r"""
-argmin(input) -> LongTensor
+argmin(input, dim=None, keepdim=False) -> LongTensor
 
-Returns the indices of the minimum value of all elements in the :attr:`input` tensor.
+Returns the indices of the minimum values of a tensor.
```
Comment (mruberry) [inline torch/_torch_docs.py:5397]: I don't think this updated statement is true.

```
torch.argmin(torch.tensor((1., 1.)))
: tensor(0)
```

If dim is None then argmin computes the flattened index of a minimal element in input. If dim is specified then I have a hard time explaining what argmin does. It's like... it computes the indices of minimal elements of... something? I don't think we can great language for this.

Looking through our internal code bases there are very few (if any) uses of this function with dim or axis (for np.argmin) set, which doesn't surprise me. You can't even use the results of this function as indices to get the minima!

```
t = torch.randn(2, 2, 2)
t
: tensor(
  [[[ 0.5317, -0.2113],
    [ 1.0522, -1.1861]],
   [[-1.8205, -0.7039],
    [-0.3392,  0.2466]]])

minima_indices = torch.argmin(t, dim=1)
minima_indices
: tensor(
  [[0, 1],
   [0, 0]])

t[minima_indices]
: 
tensor(
  [[[[ 0.5317, -0.2113],
     [ 0.5317,  0.5317]],
    [[ 1.0522, -1.1861],
     [ 1.0522,  1.0522]]],
   [[[-1.8205, -0.7039],
     [-1.8205, -1.8205]],
    [[-0.3392,  0.2466],
     [-0.3392, -0.3392]]]])
```

File: torch/_torch_docs.py [hunk 4]
```diff
@@ -7781,6 +7773,8 @@ def merge_dicts(*dicts):
 
 Args:
     {input}
+
+Keyword args:
     unbiased (bool): whether to use the unbiased estimation or not
```
Comment (mruberry) [inline torch/_torch_docs.py:7778]: This would still be in Args, unfortunately.

**Comment by mruberry** [comment]
> Note: it looks like the `Keyword Arguments` in some cases was/is meant as "this is a keyword-**only** argument". That's a weird thing to separate out, and is already indicated by the `*` in the function signatures. So I interpreted "keyword arguments" as what it says - all keyword args.

Unfortunately I think our standard is that "Keyword Arguments" = "Keyword-only Arguments"

File: torch/_torch_docs.py (hunk 3 above)
Comment (rgommers) [reply torch/_torch_docs.py:5397]: > The NumPy documentation says "along an axis," which I guess is OK? I wonder if "along an axis" is ever formally defined?

So change to _"Returns the indices of the minimum values along a dimension"_? I'm not sure there's a formal definition of "along an axis" anywhere, it's usually made clear by examples.



> The NumPy documentation correctly pairs this function with take_along_axis, which PyTorch doesn't have (but should!).

Agreed. Tracked already in gh-38349

Comment (mruberry) [reply torch/_torch_docs.py:5397]: That seems OK but we should be clear that it returns the index of *a* minimal value. That is, if a tensor has multiple minimal values then this could return the flattened index of *any* of them, and will still only return one flattened index.

Comment (mruberry) [reply torch/_torch_docs.py:5397]: So like... "Computes the flattened index of a minimal value of input when dim is None, or the indices of a minimal value along each dimension when dim is specified"?

Something about that sentence still seems a little off. It's hard to articulate that when dim is specified then multiple minimal values are returned but still just one "along each axis" is picked.

File: torch/_torch_docs.py (hunk 1 above)
Comment (rgommers) [reply torch/_torch_docs.py:2909]: Okay, the `Keyword Arguments` name is coming from Napoleon: https://github.com/sphinx-contrib/napoleon/blob/83bf1963096490dd666f93ef5a9ed1cb229fc3ec/sphinxcontrib/napoleon/docstring.py#L673. Nowhere does Napoleon make clear whether that's meant as keyworl-only yes or no. The other odd thing is that Napoleon implements two docstring styles (Google or NumPy), and neither has `Keyword Arguments` as a separate section. Google just has `Args` and NumPy only `Parameters`.

Napoleon introduced it in [this commit](https://github.com/sphinx-contrib/napoleon/commit/43423d4834ce), which points to the Khan Academy style guide. But that style guide now follows Google style: https://github.com/Khan/style-guides/blob/master/style/python.md#docstrings.

tl;dr best to get rid of `Keyword Arguments` completely.

Comment (rgommers) [reply torch/_torch_docs.py:2909]: Filed an issue for it at https://github.com/sphinx-contrib/napoleon/issues/29

**Comment by rgommers** [comment]
> Unfortunately I think our standard is that "Keyword Arguments" = "Keyword-only Arguments"

I'll revert some changes here then. Long-term I recommend getting rid of it, because it's very weird.

File: torch/_torch_docs.py (hunk 3 above)
Comment (rgommers) [reply torch/_torch_docs.py:5397]: Note that there's already a prominent note right below this sentence about duplicates:

image

The first sentence of a docstring shouldn't be too long, since it goes into the autosummary table. Capturing corner cases in that first sentence isn't necessary. So `the value` -> `a value` should do it.

**Comment by rgommers** [comment]
Updated, much shorter diff to review now.

File: torch/_torch_docs.py (hunk 1 above)
Comment (mruberry) [reply torch/_torch_docs.py]: Want to also file an issue here? We could remove that section ourselves, too.

Do you really think it's not useful to separate kwarg-only arguments, though?

Comment (rgommers) [reply torch/_torch_docs.py]: > Do you really think it's not useful to separate kwarg-only arguments, though?

It confused me, and it serves no real purpose as far as I can tell - signatures contain the same info, plus non-only-named-params should typically also be used as keywords. And it's unusual, I have never seen it elsewhere.

On the other hand, I don't feel strongly enough about it to want to make changes that touch the whole code base at once.

It would be easier if we could rename the section to `Keyword-only Arguments` for clarity, but unfortunately that's difficult.

Maybe let's see what the Napoleon folks say first.

**Review by mruberry — APPROVED** [review]
Cool!

**Comment by mruberry** [comment]
While the Github system can merge this OK, for some reason our internal tooling is failing to. Would you rebase this, @rgommers?

**Comment by rgommers** [comment]
Sure, done.
